src/backend/distributed/worker/worker_partition_protocol.c

```diff
@@ -221,6 +245,39 @@ worker_hash_partition_table(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * SyntheticShardIntervalArrayForShardMinValues returns a shard interval pointer array
+ * which gets the shardMinValues from the input shardMinValues array. Note that
+ * we simply calculate shard max values by decrementing previous shard min values by one.
+ *
+ * The function only fills the min/max values of shard the intervals. Thus, should
+ * not be used for general purpose operations.
+ */
```

Comment:
Probably the right approach here would be to introduce some intermediate data structure that `SearchCachedShardInterval` works on. If someone makes changes to `SearchCachedShardInterval` then he's unlikely to know that it gets called in a weird way.

Maybe a nice way to do this would be if ShardInterval contained a sub-struct (e.g. ShardRange) that contained `minValue` and `maxValue`. Then you could quite efficiently build an array of pointers to all the `ShardRange`s inside the `ShardInterval`s during shard pruning, and construct an array of `ShardRange`s here.

We probably don't have the time to think about mitigating potential performance implications, so this approach seems fine for now.